### DISCUSSION TITLE
How to keep using a color without it changing color when using the default tool?

### DISCUSSION THREAD

**Discussion by vladQll** [xournalpp/xournalpp#4608] [Q&A]
Hello, since I work with a wacom graphic tablet, I have configured the buttons on the tablet for certain actions one of those buttons is assigned to my default tool (pencil), and as you know the default tool is configured in the settings then it is also assigned a color, in my case I assigned the color black because it is the standard for writing, but unfortunately in many occasions I have to use other colors to write or to draw the forms geometric, but I would like that after drawing the forms geometric when returning to the predetermined tool this one continues using the color that I am using the moment and it does not change it to the one that I assign in the settings (black) I have some functions that I copy:
```app.registerUi({["menu"] = "Cycle through tools", ["callback"] = "tool", ["accelerator"] = "<Alt>s"});```

```
local toolList = {"PEN", "HIGHLIGHTER"}
local currentTool = 1

```
```
function tool()
  currentTool = currentTool % #toolList + 1
  if (toolList[currentTool] == "SELECTION") then
    app.uiAction({["action"] = "ACTION_TOOL_SELECT_" .. selectList[currentSelect]})
    print("ACTION_TOOL_SELECT_" .. selectList[currentSelect])
  else   
    app.uiAction({["action"] = "ACTION_TOOL_" .. toolList[currentTool]})
    print("ACTION_TOOL_" .. toolList[currentTool])
  end
  if (toolList[currentTool] == "PEN") then
     getColor = app.getToolInfo("active")["color"]
    app.uiAction({["action"] = "ACTION_TOOL_DEFAULT"})
    app.uiAction({["action"] = "ACTION_TOOL_LINE_STYLE_PLAIN"})
  end
end
```

@rolandlo I was told that I could obtain the color with `app.getToolInfo("active")["color"]` but I don't know how or in which part of the function I should place it. https://github.com/xournalpp/xournalpp/discussions/4564#discussioncomment-4608614

**Comment by rolandlo** [comment]
How exactly did you map one button to the default tool? Via the keyboard shortcut CTRL+SHIFT+D or in some other way?

**Comment by vladQll** [comment]
well at first I needed to press a button to go through a list of tools, the need at that time was that I wanted to change quickly from a line, to an arrow, to double arrow etc, then I copied some code snippets from some old comments, and these are in the plugin ShorcutsPlugin which is located in `C:/programFiles/xournal++/share/xournalapp/plugins` then this need I extended it and assigned it to another function to go between the highlighter and the pencil , but the problem with the pencil was that if you previously used some kind of line style exmplo dotted, when you use other tools and go back to the pencil it still had the dotted line style, that is why one of those functions runs {highlighter and default tool}, then in that same function is assigned to a combination of keys `Alt s` as we see in the following image
![image](https://user-images.githubusercontent.com/28522011/219828961-94105269-7950-4f7a-b29b-d87ecfd5f9ca.png)
 and this is the code one is for the list and the other for the function that does the magic I think:
![image](https://user-images.githubusercontent.com/28522011/219829006-6eee3823-60e8-4e74-8e48-beabc42d5abb.png)
![image](https://user-images.githubusercontent.com/28522011/219829032-7713f090-3511-444b-bf0f-f32362b32c58.png)

and subsequently for that function to be called I must link (`Alt s`) to a button in this case I did it to the button on the tablet
![image](https://user-images.githubusercontent.com/28522011/219829117-8fd01870-b717-4fec-97a9-4eaee8622f80.png)

![image](https://user-images.githubusercontent.com/28522011/219829104-f1001e62-c3e3-40f6-8711-0bdaa7f10b1a.png)

and of course apart from all that you have to configure which tool will be the default in this case it is a pencil and you also have to choose your favorite **COLOR** among other features.
![image](https://user-images.githubusercontent.com/28522011/219829170-eb1667e1-4246-4b6e-a10a-a2e17a3eecd9.png)


The problem with this is that I thought it would be the solution and that it would allow me to perform more naturally and accurately while writing fast, taking notes, making graphics with the wacom pencil but no, for example if I am using a green color and I am writing everything in green , drawing graphics, arrows, figures and I want to switch to pencil to be able to write take notes etc, this button assigned to `Alt s` will help me to switch back to pencil but not with the **Green Color** but with the color selected in the tool settings (Black).

Then I came up with another solution, I tried to ask for help or information in another discussion created for the following, based on the fact that when you install xournal for the first time and you use a tool like arrow line, etc. you select with the mouse click or pencil, the tool and if you do not want to use it, you go again with the mouse and deselect it by clicking and wonderfully you return to the pencil with the color you have been using at that time. then I started to think how to write in code that behavior so that when I use an arrow or line and press a key or better yet remembering that the arrows, lines, double arrows I have them in a function assigned to a key `alt d` then when I go through the tools and arrive to the last tool and if I press again the key this deselection as if I stopped using it and go to the pencil.

**Comment by rolandlo** [comment]
Ok, I understand better now. Here is the code to switch to normal pen with the active color:

```lua
function initUi()
app.registerUi({["menu"] = "Normal pen with active color", ["callback"] = "tool", ["accelerator"] = "<Alt>s"})
end

function tool()
  activeColor = app.getToolInfo("active")["color"]
  app.uiAction({["action"] = "ACTION_TOOL_PEN"})
  app.uiAction({["action"] = "ACTION_TOOL_LINE_STYLE_PLAIN"})
  app.uiAction({["action"] = "ACTION_SIZE_MEDIUM"})
  app.uiAction({["action"] = "ACTION_SHAPE_RECOGNIZER", ["enabled"] = false})   -- change false to true if you want shapes recognized
  app.uiAction({["action"] = "ACTION_TOOL_FILL", ["enabled"] = false})
  app.changeToolColor({["color"] = activeColor, ["tool"] = "PEN"})
end
```

It doesn't use the default tool settings. Does that help?
